File: lib/orphans.js
'use strict';

const { isObject } = require('./util');

const SCRIPT_IDS = new Set();
const START_SCRIPTS = [];
const OBJECTS = {};
const USED_OBJECTS = new Set();
const ENCHANTMENTS = new Set();
const USED_ENCHANTMENTS = new Set();
const JOURNALS = {};
const USED_JOURNALS = {};

module.exports = {
	onRecord(record, mode, recordId) {
		if(mode === 'TD') {
			return;
		}
		if(record.type === 'Script') {
			SCRIPT_IDS.add(recordId);
		} else if(record.type === 'Enchantment') {
			ENCHANTMENTS.add(recordId);
		} else if(record.type === 'Dialogue') {
			if(record.dialogue_type === 'Journal') {
				JOURNALS[recordId] = new Set();
			}
		} else {
			if(record.script) {
				SCRIPT_IDS.delete(record.script.toLowerCase());
			}
			if(record.enchanting) {
				USED_ENCHANTMENTS.add(record.enchanting.toLowerCase());
			}
			if(isObject(record)) {
				OBJECTS[recordId] = record;
			}
		}
	},
	onInfo(record, topic) {
		if(topic.dialogue_type === 'Journal' && !('quest_name' in record)) {
			JOURNALS[topic.id.toLowerCase()].add(record.data.disposition);
		}
	},
	onCellRef(record, reference, id) {
		delete OBJECTS[id];
	},
	onLevelled(record, entry, id) {
		USED_OBJECTS.add(id);
	},
	onInventory(record, entry, id) {
		USED_OBJECTS.add(id);
	},
	onScriptLine(record, line) {
		if(!line) {
			return;
		}
		const startScript = /^([,\s]*|.*?->[,\s]*)startscript[,\s]+("[^"]+"|[^,\s]+)[,\s]*$/.exec(line);
		if(startScript) {
			const [, , id] = startScript;
			return START_SCRIPTS.push(id.replace(/"/g, ''));
		}
		const first = /^([,\s]*|.*?->[,\s]*)(placeatme|addsoulgem|additem|equip|drop|placeatpc|placeitemcell|placeitem)[,\s]+(?:("[^"]+"?)(?:.*)|([^,\s"]+)(?:[,\s]+|$))/.exec(line)
			?? /^([,\s]*|.*?->[,\s]*)(journal|setjournalindex)[,\s]+(?:("[^"]+"?)|([^,\s"]+))[,\s]+([\d]+)/.exec(line);
		if(first) {
			const [, , command, quotedId, unquotedId, argument] = first;
			let id = unquotedId;
			if(quotedId) {
				id = quotedId.replace(/"/g, '');
			}
			if(command === 'journal' || command === 'setjournalindex') {
				if(!USED_JOURNALS[id]) {
					USED_JOURNALS[id] = [];
				}
				return USED_JOURNALS[id].push(+argument);
			}
			return USED_OBJECTS.add(id);
		}
		const second = /^([,\s]*|.*?->[,\s]*)(addtolevcreature|addtolevitem)[,\s]+("[^"]+"|[^,\s]+)[,\s]+("[^"]+"|[^,\s]+)([,\s]+|$)/.exec(line);
		if(second) {
			const [, , , , id] = second;
			return USED_OBJECTS.add(id.replace(/"/g, ''));
		}
	},
	onEnd() {
		START_SCRIPTS.forEach(id => {
			SCRIPT_IDS.delete(id);
		});
		SCRIPT_IDS.forEach(id => {
			console.log('Script', id, 'is never started');
		});
		ENCHANTMENTS.forEach(id => {
			if(!USED_ENCHANTMENTS.has(id)) {
				console.log('Enchantment', id, 'is not used');
			}
		});
		USED_OBJECTS.forEach(id => {
			delete OBJECTS[id];
		});
		Object.values(OBJECTS).forEach(record => {
			console.log(record.type, record.id, 'is not used');
		});
		for(const id in JOURNALS) {
			const used = USED_JOURNALS[id];
			if(used) {
				JOURNALS[id].forEach(index => {
					if(!used.includes(index)) {
						console.log('Journal index', index, 'in', id, 'is unused');
					}
				});
			} else {
				console.log('Journal', id, 'is not used');
			}
		}
	}
};

File: src/validators/orphans.rs
use std::collections::{HashMap, HashSet};

use super::Context;
use crate::{context::Mode, handlers::Handler, util::update_or_insert};
use regex::{Error, Regex};
use tes3::esp::{
    Cell, Dialogue, DialogueInfo, DialogueType2, EditorId, FixedString, QuestState, Reference,
    TES3Object, TypeInfo,
};

pub struct OrphanValidator {
    script_ids: HashSet<String>,
    start_scripts: Vec<String>,
    objects: HashMap<String, &'static str>,
    used_objects: HashSet<String>,
    enchantments: HashSet<String>,
    used_enchantments: HashSet<String>,
    journals: HashMap<String, HashSet<i32>>,
    used_journals: HashMap<String, Vec<i32>>,
    startscript: Regex,
    firstarg: Regex,
    journal: Regex,
    secondarg: Regex,
}

fn is_journal(dialogue: &Dialogue) -> bool {
    dialogue.dialogue_type == DialogueType2::Journal
}

impl Handler<'_> for OrphanValidator {
    fn on_record(&mut self, context: &Context, record: &TES3Object) {
        if context.mode == Mode::TD {
            return;
        }
        match record {
            TES3Object::Dialogue(dialogue) => {
                if is_journal(dialogue) {
                    self.journals.insert(
                        record.editor_id_ascii_lowercase().into_owned(),
                        HashSet::new(),
                    );
                }
            }
            TES3Object::Enchanting(_) => {
                self.enchantments
                    .insert(record.editor_id_ascii_lowercase().into_owned());
            }
            TES3Object::Script(_) => {
                self.script_ids
                    .insert(record.editor_id_ascii_lowercase().into_owned());
            }
            TES3Object::Activator(r) => {
                self.remove_script(&r.script);
                self.insert_object(record);
            }
            TES3Object::Alchemy(r) => {
                self.remove_script(&r.script);
                self.insert_object(record);
            }
            TES3Object::Apparatus(r) => {
                self.remove_script(&r.script);
                self.insert_object(record);
            }
            TES3Object::Armor(r) => {
                self.remove_script(&r.script);
                self.insert_object(record);
                self.add_enchantment(&r.enchanting);
            }
            TES3Object::Book(r) => {
                self.remove_script(&r.script);
                self.insert_object(record);
                self.add_enchantment(&r.enchanting);
            }
            TES3Object::Clothing(r) => {
                self.remove_script(&r.script);
                self.insert_object(record);
                self.add_enchantment(&r.enchanting);
            }
            TES3Object::Container(r) => {
                self.remove_script(&r.script);
                self.insert_object(record);
            }
            TES3Object::Creature(r) => {
                self.remove_script(&r.script);
                self.insert_object(record);
            }
            TES3Object::Door(r) => {
                self.remove_script(&r.script);
                self.insert_object(record);
            }
            TES3Object::Ingredient(r) => {
                self.remove_script(&r.script);
                self.insert_object(record);
            }
            TES3Object::LeveledCreature(_) => {
                self.insert_object(record);
            }
            TES3Object::LeveledItem(_) => {
                self.insert_object(record);
            }
            TES3Object::Light(r) => {
                self.remove_script(&r.script);
                self.insert_object(record);
            }
            TES3Object::Lockpick(r) => {
                self.remove_script(&r.script);
                self.insert_object(record);
            }
            TES3Object::MiscItem(r) => {
                self.remove_script(&r.script);
                self.insert_object(record);
            }
            TES3Object::Npc(r) => {
                self.remove_script(&r.script);
                self.insert_object(record);
            }
            TES3Object::Probe(r) => {
                self.remove_script(&r.script);
                self.insert_object(record);
            }
            TES3Object::RepairItem(r) => {
                self.remove_script(&r.script);
                self.insert_object(record);
            }
            TES3Object::StartScript(r) => {
                self.remove_script(&r.script);
            }
            TES3Object::Static(_) => {
                self.insert_object(record);
            }
            TES3Object::Weapon(r) => {
                self.remove_script(&r.script);
                self.insert_object(record);
                self.add_enchantment(&r.enchanting);
            }
            _ => {}
        }
    }

    fn on_info(&mut self, _: &Context, record: &DialogueInfo, topic: &Dialogue) {
        if is_journal(topic) && record.quest_state != Some(QuestState::Name) {
            if let Some(indices) = self.journals.get_mut(&topic.id.to_ascii_lowercase()) {
                indices.insert(record.data.disposition);
            }
        }
    }

    fn on_cellref(
        &mut self,
        _: &Context,
        _: &Cell,
        _: &Reference,
        id: &str,
        _: &[&Reference],
        _: usize,
    ) {
        self.objects.remove(id);
    }

    fn on_leveled(&mut self, _: &Context, _: &TES3Object, entry: &(String, u16)) {
        self.used_objects.insert(entry.0.to_ascii_lowercase());
    }

    fn on_inventory(&mut self, _: &Context, _: &TES3Object, entry: &(i32, FixedString<32>)) {
        self.used_objects.insert(entry.1.to_ascii_lowercase());
    }

    fn on_scriptline(
        &mut self,
        _: &Context,
        _: &TES3Object,
        code: &str,
        _: &str,
        _: &Dialogue,
        _: &str,
    ) {
        if code.is_empty() {
            return;
        }
        if let Some(captures) = self.startscript.captures(code) {
            let id = captures.get(2).unwrap().as_str();
            self.start_scripts.push(id.replace('"', ""));
        } else if let Some(captures) = self.firstarg.captures(code) {
            if let Some(quoted) = captures.get(3) {
                self.used_objects.insert(quoted.as_str().replace('"', ""));
            } else if let Some(unquoted) = captures.get(4) {
                self.used_objects.insert(unquoted.as_str().to_string());
            }
        } else if let Some(captures) = self.journal.captures(code) {
            let id;
            if let Some(quoted) = captures.get(3) {
                id = quoted.as_str().replace('"', "");
            } else if let Some(unquoted) = captures.get(4) {
                id = unquoted.as_str().to_string();
            } else {
                return;
            }
            if let Ok(index) = captures.get(5).unwrap().as_str().parse::<i32>() {
                update_or_insert(&mut self.used_journals, id, |e| e.push(index));
            }
        } else if let Some(captures) = self.secondarg.captures(code) {
            self.used_objects
                .insert(captures.get(4).unwrap().as_str().replace('"', ""));
        }
    }

    fn on_end(&mut self, _: &Context) {
        for id in &self.start_scripts {
            self.script_ids.remove(id);
        }
        for id in &self.script_ids {
            println!("Script {} is never started", id);
        }
        for id in &self.enchantments {
            if !self.used_enchantments.contains(id) {
                println!("Enchantment {} is not used", id);
            }
        }
        for id in &self.used_objects {
            self.objects.remove(id);
        }
        for (id, typename) in &self.objects {
            println!("{} {} is not used", typename, id);
        }
        for (id, indices) in &self.journals {
            if let Some(used) = self.used_journals.get(id) {
                for index in indices {
                    if !used.contains(index) {
                        println!("Journal index {} in {} is unused", index, id);
                    }
                }
            } else {
                println!("Journal {} is not used", id);
            }
        }
    }
}

impl OrphanValidator {
    pub fn new() -> Result<Self, Error> {
        let startscript =
            Regex::new(r#"^([,\s]*|.*?->[,\s]*)startscript[,\s]+("[^"]+"|[^,\s]+)[,\s]*$"#)?;
        let firstarg = Regex::new(
            r#"^([,\s]*|.*?->[,\s]*)(placeatme|addsoulgem|additem|equip|drop|placeatpc|placeitemcell|placeitem)[,\s]+(?:("[^"]+"?)(?:.*)|([^,\s"]+)(?:[,\s]+|$))"#,
        )?;
        let journal = Regex::new(
            r#"^([,\s]*|.*?->[,\s]*)(journal|setjournalindex)[,\s]+(?:("[^"]+"?)|([^,\s"]+))[,\s]+([\d]+)"#,
        )?;
        let secondarg = Regex::new(
            r#"^([,\s]*|.*?->[,\s]*)(addtolevcreature|addtolevitem)[,\s]+("[^"]+"|[^,\s]+)[,\s]+("[^"]+"|[^,\s]+)([,\s]+|$)"#,
        )?;
        Ok(Self {
            script_ids: HashSet::new(),
            start_scripts: Vec::new(),
            objects: HashMap::new(),
            used_objects: HashSet::new(),
            enchantments: HashSet::new(),
            used_enchantments: HashSet::new(),
            journals: HashMap::new(),
            used_journals: HashMap::new(),
            startscript,
            firstarg,
            journal,
            secondarg,
        })
    }

    fn insert_object(&mut self, record: &TES3Object) {
        self.objects.insert(
            record.editor_id_ascii_lowercase().to_string(),
            record.type_name(),
        );
    }

    fn remove_script(&mut self, script: &str) {
        if !script.is_empty() {
            self.script_ids.remove(&script.to_ascii_lowercase());
        }
    }

    fn add_enchantment(&mut self, enchantment: &str) {
        if !enchantment.is_empty() {
            self.used_enchantments
                .insert(enchantment.to_ascii_lowercase());
        }
    }
}
